packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+// @ts-check
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        const handler = wasFulfilled ? watcher.onFulfilled : watcher.onRejected;
+        if (handler) {
+          handler(value, ...args);
+        }
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    if (deadPromisesRaw) {
```

Comment:
Since the `'deadPromises'` (not `deadPromisesRaw`, that's just a variable name) entry in the vatstore is created by the shutdown operation, you're correct that it will be missing at vat creation time and should always exist at startup time thereafter.  Consequently the existence of the `'deadPromises'` key in the vatstore is a sort of proxy for "yes, we are doing a restart".  Its absence on restart would be symptom that something terrible had gone wrong with the shutdown, and I suspect in that case restoring the promise watchers would be the least of our concerns.

That said, I like the idea of decoupling the fetching of the list of dead promises from the restoration of the promise watchers.  I think the approach I'd take would, rather than changing the conditional on the loop, is to simply initialize the `deadPromises` variable to an empty Set in the case where the `'deadPromises'` key is absent.